autotest/ogr/ogr_pg.py

```diff
@@ -915,126 +1127,51 @@ def test_ogr_pg_16():
-        (
-            "MULTILINESTRING ((10 10 1 2,20 20 3 4),(30 30 5 6,40 40 7 8))",
-            "MULTILINESTRING ZM ((10 10 1 2,20 20 3 4),(30 30 5 6,40 40 7 8))",
-        ),
-        (
-            "MULTIPOLYGON(((0 0 0 1,4 0 0 1,4 4 0 1,0 4 0 1,0 0 0 1),(1 1 0 5,2 1 0 5,2 2 0 5,1 2 0 5,1 1 0 5)),((-1 -1 0 10,-1 -2 0 10,-2 -2 0 10,-2 -1 0 10,-1 -1 0 10)))",
-            "MULTIPOLYGON ZM (((0 0 0 1,4 0 0 1,4 4 0 1,0 4 0 1,0 0 0 1),(1 1 0 5,2 1 0 5,2 2 0 5,1 2 0 5,1 1 0 5)),((-1 -1 0 10,-1 -2 0 10,-2 -2 0 10,-2 -1 0 10,-1 -1 0 10)))",
-        ),
-        (
-            "GEOMETRYCOLLECTION(POINT(2 3 11 101),LINESTRING(2 3 12 102,3 4 13 103))",
-            "GEOMETRYCOLLECTION ZM (POINT ZM (2 3 11 101),LINESTRING ZM (2 3 12 102,3 4 13 103))",
-        ),
-        (
-            "TRIANGLE ((0 0 0 0,100 0 100 1,0 100 100 0,0 0 0 0))",
-            "TRIANGLE ZM ((0 0 0 0,100 0 100 1,0 100 100 0,0 0 0 0))",
-        ),
-        (
-            "TIN (((0 0 0 0,0 0 1 0,0 1 0 0,0 0 0 0)),((0 0 0 0,0 1 0 0,1 1 0 0,0 0 0 0)))",
-            "TIN ZM (((0 0 0 0,0 0 1 0,0 1 0 0,0 0 0 0)),((0 0 0 0,0 1 0 0,1 1 0 0,0 0 0 0)))",
-        ),
-        (
-            "POLYHEDRALSURFACE (((0 0 0 0,0 0 1 0,0 1 1 0,0 1 0 0,0 0 0 0)),((0 0 0 0,0 1 0 0,1 1 0 0,1 0 0 0,0 0 0 0)),((0 0 0 0,1 0 0 0,1 0 1 0,0 0 1 0,0 0 0 0)),((1 1 0 0,1 1 1 0,1 0 1 0,1 0 0 0,1 1 0 0)),((0 1 0 0,0 1 1 0,1 1 1 0,1 1 0 0,0 1 0 0)),((0 0 1 0,1 0 1 0,1 1 1 0,0 1 1 0,0 0 1 0)))",
-            "POLYHEDRALSURFACE ZM (((0 0 0 0,0 0 1 0,0 1 1 0,0 1 0 0,0 0 0 0)),((0 0 0 0,0 1 0 0,1 1 0 0,1 0 0 0,0 0 0 0)),((0 0 0 0,1 0 0 0,1 0 1 0,0 0 1 0,0 0 0 0)),((1 1 0 0,1 1 1 0,1 0 1 0,1 0 0 0,1 1 0 0)),((0 1 0 0,0 1 1 0,1 1 1 0,1 1 0 0,0 1 0 0)),((0 0 1 0,1 0 1 0,1 1 1 0,0 1 1 0,0 0 1 0)))",
-        ),
-    )
-
-    # This layer is also used in ogr_pg_21() test.
-    gdaltest.pg_ds.ExecuteSQL("CREATE TABLE testgeom (ogc_fid integer)")
-
-    # XXX - mloskot - if 'public' is omitted, then OGRPGDataSource::DeleteLayer fails, line 438
-    sql_lyr = gdaltest.pg_ds.ExecuteSQL(
-        "SELECT AddGeometryColumn('public','testgeom','wkb_geometry',-1,'GEOMETRY',4)"
-    )
-    gdaltest.pg_ds.ReleaseResultSet(sql_lyr)
-    for i, geom in enumerate(geometries):
-        gdaltest.pg_ds.ExecuteSQL(
-            "INSERT INTO testgeom (ogc_fid,wkb_geometry) \
-                                    VALUES (%d,GeomFromEWKT('%s'))"
-            % (i, geom[0])
-        )
-
-    # We need to re-read layers
-    gdaltest.pg_ds.Destroy()
-    gdaltest.pg_ds = ogr.Open("PG:" + gdaltest.pg_connection_string, update=1)
-    assert gdaltest.pg_ds is not None, "can not re-open datasource"
-
-    #
-    # Test reading 4-dim geometries normalized to OGC WKT form.
-    #
+@only_with_postgis
+@pytest.mark.usefixtures("testgeom")
+def test_ogr_pg_20(pg_ds):
 
-    layer = gdaltest.pg_ds.GetLayerByName("testgeom")
+    layer = pg_ds.GetLayerByName("testgeom")
```

Comment:
The test currently runs with "ERROR 6: SetFeature : unsupported operation on a read-only datasource." errors, that happens to be undetected:

```suggestion
    ds = reconnect(pg_ds, update=1)
    layer = ds.GetLayerByName("testgeom")
```

The PG driver is quite lax about checking the update mode, so most of this issues are undetected. I also see that test_ogr_22 would also need an updatable dataset, although it doesn't hit any code path that checks permission. And probably other tests. Perhaps there should be a pg_updatable_ds fixture, and use it more extensively?

Reply:
Having `pg_ds` be read-only is an oversight (for example, `mysql_ds` opens in update mode). Since `pg_ds` is scoped to each test, is there a use for a read-only version?

Reply:
>  is there a use for a read-only version?

probably not, unless we test read-only checks, which we probably don't do in ogr_pg

Reply:
It depends on what the tests are testing. In managed production systems users tend to have minimum rights. For example Mapserver user having a layer through OGR connection should not have any other rights than SELECT on the data table, geometry_columns, spatial_ref_sys and maybe some other metadata and system tables. Developers tend to forget that there are users who operate with limited rights. However, I have not seen any reports about GDAL failing because of this.
I guess that opening a connection as read-only is not the same than opening a connection as a user who has only SELECT rights. Do we test the latter in any place? Is there any need for such tests, for example try to update without proper rights?

Reply:
There are 2 things regarding rights: the update flag that is handled by GDAL (at the driver level, and the PG driver lacks checks to validate that operations that update are done with the update flag set), and possibly opening the underlying datasource in read-only (which the PG driver does not currently either. According to https://stackoverflow.com/questions/41960217/connecting-to-postgres-database-with-read-only-access,it can be done at the session level with "SET SESSION CHARACTERISTICS AS TRANSACTION READ ONLY;").
So currently opening a PG connection in read-only mode does not prevent a user from modifying the database, if he has rights at the PostgreSQL level. Which is kind of unexpected.